docs/architecture/adr-022-abci-errors.md

```diff
@@ -0,0 +1,60 @@
+# ADR 023: ABCI Codespaces
+
+## Changelog
+
+- *2018-09-01* Initial version
+
+## Context
+
+ABCI errors should provide an abstraction between application details
+and the client interface responsible for formatting & displaying errors to the user.
+
+Currently, this abstraction consists of a single integer (the `code`), where any
+`code > 0` is considered an error (ie. invalid transaction) and all type
+information about the error is contained in the code. This integer is
+expected to be decoded by the client into a known error string, where any
+more specific data is contained in the `data`.
+
+In a [previous conversation](https://github.com/tendermint/abci/issues/165#issuecomment-353704015),
+it was suggested that not all non-zero codes need to be errors, hence why it's called `code` and not `error code`.
+It is unclear exactly how the semantics of the `code` field will evolve, though
+better lite-client proofs (like discussed for tags
+[here](https://github.com/tendermint/tendermint/issues/1007#issuecomment-413917763))
+may play a role.
+
+For now, note that having all type information in a single integer
+places an unfortunate requirement on all users of the `code` system to
+co-ordinate up-front on which `code` means what, hence restricting their ability experiment
```

Comment:
*ability to experiment*

That's one problem, but the other problem is that it precludes an easy coordination method between "module implementers" and "client implementers". With an unbounded error domain (such as a string), module implementers can pick a globally unique prefix & error code set, so client implementers could easily implement support for "module A" regardless of which Hub/zone it was running in and which other modules were running with it. With only error codes, globally unique codes are difficult/impossible, as the space is finite and collisions are likely without an easy way to coordinate.

Reply:
added this verbatim :)